ui/framework/src/ui-framework/imodel-components/models-tree/ModelsTree.tsx

```diff
@@ -157,25 +166,50 @@ export function ModelsTree(props: ModelsTreeProps) {
 export const IModelConnectedModelsTree = connectIModelConnection(null, null)(ModelsTree); // eslint-disable-line @typescript-eslint/naming-convention
 
 function useModelsTreeNodeLoader(props: ModelsTreeProps) {
+  // note: this is a temporary workaround for auto-update not working on ruleset variable changes - instead
+  // of auto-updating we just re-create the node loader by re-creating the ruleset
+  const rulesets = React.useMemo(() => {
+    return {
+      general: (!props.enableElementsClassGrouping) ? { ...RULESET_MODELS } : /* istanbul ignore next */ { ...RULESET_MODELS_GROUPED_BY_CLASS },
+      search: { ...RULESET_MODELS_SEARCH },
+    };
+  }, [props.filteredElementIds]); // eslint-disable-line react-hooks/exhaustive-deps
+  // const rulesets = {
+  //   general: (!props.enableElementsClassGrouping) ? RULESET_MODELS : /* istanbul ignore next */ RULESET_MODELS_GROUPED_BY_CLASS,
+  //   search: RULESET_MODELS_SEARCH,
+  // };
+
   const { nodeLoader, onItemsRendered } = usePresentationTreeNodeLoader({
     imodel: props.iModel,
     dataProvider: props.dataProvider,
-    ruleset: (!props.enableElementsClassGrouping) ? RULESET_MODELS : /* istanbul ignore next */ RULESET_MODELS_GROUPED_BY_CLASS,
+    ruleset: rulesets.general,
     appendChildrenCountForGroupingNodes: (props.enableElementsClassGrouping === ClassGroupingOption.YesWithCounts),
     pagingSize: PAGING_SIZE,
     enableHierarchyAutoUpdate: props.enableHierarchyAutoUpdate,
   });
-  const { nodeLoader: searchNodeLoader, onItemsRendered: onSeachItemsRendered } = usePresentationTreeNodeLoader({
+  const { nodeLoader: searchNodeLoader, onItemsRendered: onSearchItemsRendered } = usePresentationTreeNodeLoader({
     imodel: props.iModel,
     dataProvider: props.dataProvider,
-    ruleset: RULESET_MODELS_SEARCH,
+    ruleset: rulesets.search,
     pagingSize: PAGING_SIZE,
     enableHierarchyAutoUpdate: props.enableHierarchyAutoUpdate,
   });
 
+  const activeNodeLoader = props.filterInfo?.filter ? searchNodeLoader : nodeLoader;
+  const activeItemsRenderedCallback = props.filterInfo?.filter ? onSearchItemsRendered : onItemsRendered;
+
+  const vars = Presentation.presentation.vars(activeNodeLoader.dataProvider.rulesetId);
+  if (props.filteredElementIds) {
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    vars.setId64s("filtered-element-ids", props.filteredElementIds);
+  } else {
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    vars.unset("filtered-element-ids");
+  }

```

Comment:
Looks like this should be called only when `filteredElementIds` or `vars` changes?

Reply:
I don't think that matters much - this check is done in `RulesetVariablesManager`, so in case of same value it'll just return early without emitting an `onVariableChanged` event.